send_email_pdf/main.py

```diff
@@ -0,0 +1,293 @@
+then uploads the pdf to an s3 bucket and sends an email of the pdf."""
+
+
+import sys
+from os import environ
+from email.mime.multipart import MIMEMultipart
+from email.mime.application import MIMEApplication
+from datetime import datetime
+
+from dotenv import load_dotenv
+from psycopg2.extensions import connection
+from psycopg2 import connect, Error
+import pandas as pd
+import plotly.graph_objects as go
+from xhtml2pdf import pisa
+from boto3 import client
+
+
+PDF_FILE_NAME = "Media-Sentiment.pdf"
+PDF_FILE_PATH = "/tmp/Media-Sentiment.pdf"
+
+
+def join_all_stories_info(conn: connection) -> pd.DataFrame:   # pragma: no cover
+    "Function that joins the stories and story sources SQL tables"
+    query = """SELECT story_id, title, description,url, pub_date, media_sentiment,
+    source_name FROM stories JOIN sources ON stories.source_id = sources.source_id"""
+    with conn.cursor() as cur:
+        cur.execute(query)
+        tuples_list = cur.fetchall()
+    stories_df = pd.DataFrame(tuples_list, columns=[
+        "story_id", "title", "description", "url", "pub_date",
+        "article_sentiment", "source_name"])
+    return stories_df
+
+
+def join_all_reddit_info(conn: connection) -> pd.DataFrame:   # pragma: no cover
+    "Function that joins all reddit sources SQL tables"
+    query = """SELECT * FROM reddit_article"""
+    with conn.cursor() as cur:
+        cur.execute(query)
+        tuples_list = cur.fetchall()
+    columns_list = ["article_id", "domain", "title", "article_url",
+                    "url", "re_sentiment_mean", "re_sentiment_st_dev",
+                    "re_sentiment_median", "re_vote_score", "re_upvote_ratio",
+                    "re_post_comments", "re_processed_comments", "re_created_timestamp"]
+    reddit_df = pd.DataFrame(tuples_list, columns=columns_list)
+    return reddit_df
+
+
+def get_db_connection():   # pragma: no cover
+    """Establishes a connection with the PostgreSQL database."""
+    try:
+        conn = connect(
+            user=environ.get("DATABASE_USERNAME"),
+            password=environ.get("DATABASE_PASSWORD"),
+            host=environ.get("DATABASE_IP"),
+            port=environ.get("DATABASE_PORT"),
+            database=environ.get("DATABASE_NAME"),)
+        print("Database connection established successfully.")
+        return conn
+    except Error as err:
+        print("Error connecting to database: ", err)
+        sys.exit()
+
+
+def get_titles(titles) -> str:
+    """Function that takes out all the title names from the top stories
+    and returns them as a html ready string"""
+    title_str = "<ul>"
+    for title in titles:
+        title_str += f"<li>{title}</li>"
+    title_str += "</ul>"
+    return title_str
+
+
+def create_report(stories_data: pd.DataFrame, reddit_data: pd.DataFrame) -> str:
+    """Creates the HTML template for the report, including all visualizations as
+    images within the html wrapper"""
```

Comment:
This function is very large in size, can the creation of the figures be separated into smaller functions?